File: backend/src/main/java/com/pranay/ems/service/LeaveService.java
package com.pranay.ems.service;

import com.pranay.ems.dto.request.LeaveRequestDto;
import com.pranay.ems.dto.response.LeaveResponse;
import com.pranay.ems.enums.LeaveStatus;

import java.util.List;

public interface LeaveService {

    LeaveResponse applyLeave(LeaveRequestDto request);

    LeaveResponse approveLeave(Long leaveId, String approvedBy);

    LeaveResponse rejectLeave(Long leaveId, String approvedBy);

    List<LeaveResponse> getAllLeaves();

    LeaveResponse getLeaveById(Long id);

    List<LeaveResponse> getLeavesByEmployee(Long employeeId);

    List<LeaveResponse> getLeavesByStatus(LeaveStatus status);

}

File: frontend/src/services/leaveService.js
import api from "../api/axios";

/* ==============================
   Apply Leave
================================ */

export const applyLeave = async (leave) => {

    const response = await api.post(
        "/api/v1/leaves",
        leave
    );

    return response.data;

};

/* ==============================
   Get All Leaves
================================ */

export const getAllLeaves = async () => {

    const response = await api.get(
        "/api/v1/leaves"
    );

    return response.data;

};

/* ==============================
   Get Leave By Employee
================================ */

export const getLeavesByEmployee = async (employeeId) => {

    const response = await api.get(
        `/api/v1/leaves/employee/${employeeId}`
    );

    return response.data;

};

/* ==============================
   Approve Leave
================================ */

export const approveLeave = async (
    leaveId,
    approvedBy
) => {

    const response = await api.put(
        `/api/v1/leaves/${leaveId}/approve?approvedBy=${approvedBy}`
    );

    return response.data;

};

/* ==============================
   Reject Leave
================================ */

export const rejectLeave = async (
    leaveId,
    approvedBy
) => {

    const response = await api.put(
        `/api/v1/leaves/${leaveId}/reject?approvedBy=${approvedBy}`
    );

    return response.data;

};

/* ==============================
   Get Leave By Status
================================ */

export const getLeavesByStatus = async (status) => {

    const response = await api.get(
        `/api/v1/leaves/status/${status}`
    );

    return response.data;

};